### PR TITLE
Refactor/nfts remove cd

### PR DESCRIPTION
**Pull request by charleenfei**
<This is intended as a guide when describing-reviewing pull requests for go-centrifuge>
Issue Link: <Github link>

### CODE REVIEW

File: documents/anchor.go [hunk 1]
```diff
@@ -25,11 +24,11 @@ type updaterFunc func(id []byte, model Model) error
 // AnchorDocument add signature, requests signatures, anchors document, and sends the anchored document
 // to collaborators
 func AnchorDocument(ctx context.Context, model Model, proc AnchorProcessor, updater updaterFunc) (Model, error) {
-	cd, err := model.PackCoreDocument()
+	dm, err := model.PackCoreDocument()
 	if err != nil {
 		return nil, err
 	}
-
+	cd := dm.Document
```
Comment (mikiquantum) [inline documents/anchor.go]: too verbose? `dm.Document.CurrentVersion`

File: documents/invoice/model.go [hunk 2]
```diff
@@ -290,19 +289,25 @@ func (i *Invoice) PackCoreDocument() (*coredocumentpb.CoreDocument, error) {
 		return nil, errors.NewTypedError(err, errors.New("couldn't get InvoiceSalts"))
 	}
 
-	coreDoc := new(coredocumentpb.CoreDocument)
-	proto.Merge(coreDoc, i.CoreDocument)
-	coreDoc.EmbeddedData = &invoiceAny
-	coreDoc.EmbeddedDataSalts = documents.ConvertToProtoSalts(invoiceSalts)
-	return coreDoc, err
+	err = i.CoreDocumentModel.PackCoreDocument(&invoiceAny, documents.ConvertToProtoSalts(invoiceSalts))
```
Comment (mikiquantum) [inline documents/invoice/model.go]: I like it.

File: documents/invoice/model.go [hunk 3]
```diff
@@ -325,10 +330,11 @@ func (i *Invoice) UnpackCoreDocument(coreDoc *coredocumentpb.CoreDocument) error
 		i.InvoiceSalts = documents.ConvertToProofSalts(coreDoc.EmbeddedDataSalts)
 	}
 
-	i.CoreDocument = new(coredocumentpb.CoreDocument)
-	proto.Merge(i.CoreDocument, coreDoc)
-	i.CoreDocument.EmbeddedDataSalts = nil
-	i.CoreDocument.EmbeddedData = nil
+	i.CoreDocumentModel = new(documents.CoreDocumentModel)
```
Comment (mikiquantum) [inline documents/invoice/model.go]: Same way we did pack, we can add `coreDocumentModel.UnpackCoreDocument(coreDocModel *documents.CoreDocumentModel)`.

File: documents/invoice/service.go [hunk 4]
```diff
@@ -127,11 +125,12 @@ func (s service) Create(ctx context.Context, inv documents.Model) (documents.Mod
 		return nil, uuid.Nil, nil, err
 	}
 
-	cd, err := inv.PackCoreDocument()
+	dm, err := inv.PackCoreDocument()
 	if err != nil {
 		return nil, uuid.Nil, nil, err
 	}
 
+	cd := dm.Document
```
Comment (mikiquantum) [inline documents/invoice/service.go]: Same here, on the verbosity side.

File: documents/invoice/service.go [hunk 5]
```diff
@@ -147,11 +146,11 @@ func (s service) Update(ctx context.Context, inv documents.Model) (documents.Mod
 		return nil, uuid.Nil, nil, errors.NewTypedError(documents.ErrDocumentConfigAccountID, err)
 	}
 
-	cd, err := inv.PackCoreDocument()
+	dm, err := inv.PackCoreDocument()
 	if err != nil {
 		return nil, uuid.Nil, nil, errors.NewTypedError(documents.ErrDocumentPackingCoreDocument, err)
 	}
-
+	cd := dm.Document
```
Comment (mikiquantum) [inline documents/invoice/service.go]: same

File: documents/invoice/service.go [hunk 6]
```diff
@@ -172,11 +171,11 @@ func (s service) Update(ctx context.Context, inv documents.Model) (documents.Mod
 
 // DeriveInvoiceResponse returns create response from invoice model
 func (s service) DeriveInvoiceResponse(doc documents.Model) (*clientinvoicepb.InvoiceResponse, error) {
-	cd, err := doc.PackCoreDocument()
+	dm, err := doc.PackCoreDocument()
 	if err != nil {
 		return nil, errors.NewTypedError(documents.ErrDocumentPackingCoreDocument, err)
 	}
-
+	cd := dm.Document
```
Comment (mikiquantum) [inline documents/invoice/service.go]: same

File: documents/model.go [hunk 7]
```diff
@@ -163,18 +175,44 @@ func (m *CoreDocumentModel) PrepareNewVersion(collaborators []string) (*CoreDocu
...
 }
 
+// NewWithCollaborators generates new core document, adds collaborators, adds read rules and fills salts
+func (m *CoreDocumentModel) NewWithCollaborators(collaborators []string) (*CoreDocumentModel, error) {
+	dm := NewCoreDocModel()
+	ids, err := identity.CentIDsFromStrings(collaborators)
+	if err != nil {
+		return nil, errors.New("failed to decode collaborator: %v", err)
+	}
+	cd := dm.Document
```
Comment (mikiquantum) [inline documents/model.go]: same

File: documents/purchaseorder/model.go [hunk 8]
```diff
@@ -53,16 +52,21 @@ type PurchaseOrder struct {
...
 func (p *PurchaseOrder) ID() ([]byte, error) {
-	coreDoc, err := p.PackCoreDocument()
+	coreDocModel, err := p.PackCoreDocument()
 	if err != nil {
 		return []byte{}, err
 	}
+	if coreDocModel.Document == nil {
+		return []byte{}, errors.New("nil core document")
+	}
+	coreDoc := coreDocModel.Document
```
Comment (mikiquantum) [inline documents/purchaseorder/model.go]: verbose?

File: p2p/receiver/validator.go [hunk 9]
```diff
@@ -105,30 +101,6 @@ func HandshakeValidator(networkID uint32, idService identity.Service) ValidatorG
 	}
 }
 
-// DocumentAccessValidator validates the GetDocument request against the AccessType indicated in the request
-func DocumentAccessValidator(doc *coredocumentpb.CoreDocument, docReq *p2ppb.GetDocumentRequest, requesterCentID identity.CentID) error {
```
Comment (mikiquantum) [inline p2p/receiver/validator.go]: I think is better to comment this out knowing that we will need this soon. Or has it been moved?

Comment (charleenfei) [reply p2p/receiver/validator.go]: It has been moved to the CoreDocModel.

Comment (mikiquantum) [reply p2p/receiver/validator.go]: ok

**Pull request merged**